wallet/invoices.c

```diff
@@ -89,13 +92,82 @@ struct invoices *invoices_new(const tal_t *ctx,
 	return invs;
 }
 
+static void install_expiration_timer(struct invoices *invoices,
+				     struct invoice *i,
+				     u64 now);
+static void trigger_expiration(struct invoice *i)
+{
+	struct invoices *invoices = (struct invoices *)tal_parent(i);
+	u64 now = time_now().ts.tv_sec;
+	sqlite3_stmt *stmt;
+	struct invoice_waiter *w;
+	const tal_t *tmpctx;
+
+	/* There may be discrepancies between time_mono
+	 * (used by the timer system) and time_now (used
+	 * by the expiry time measurements). So check that
+	 * time_now is reached. */
+	if (i->expiry_time <= now) {
+		tmpctx = tal_tmpctx(i);
+
+		/* Update in-memory and db. */
+		i->state = EXPIRED;
+		stmt = db_prepare(invoices->db,
+				  "UPDATE invoices"
+				  "   SET state = ?"
+				  " WHERE id = ?;");
+		sqlite3_bind_int(stmt, 1, EXPIRED);
+		sqlite3_bind_int64(stmt, 2, i->id);
+		db_exec_prepared(invoices->db, stmt);
+		/* Wake up all waiters. */
+		while ((w = list_pop(&i->waitone_waiters,
+				     struct invoice_waiter,
+				     list)) != NULL) {
+			tal_steal(tmpctx, w);
+			trigger_invoice_waiter(w, i);
+		}
+
+		tal_free(tmpctx);
+		return;
+	} else
+		install_expiration_timer(invoices, i, now);
```

Comment:
This branch has never been tested (hard to do, I know), because AFAICT it will assert.

Suggest we make install_expiration_timer() more robust, see below.